### ISSUE TITLE
Support objects as items

### ISSUE THREAD

**Issue by jounik** [vaadin/vaadin-combo-box#96]
- A way to define the display value
- Filtering should use the display value instead of the raw item value? Custom filtering?
- Should the `value` (sent as a part of a form submission for instance) be the raw item value in the items array?

**Comment by jounik** [comment]
2d timebox for research. @jounik will provide use cases before starting the research.

**Comment by vaadin-marcus** [comment]
I think you should be able to also configure the value sent in forms. Most likely you don't want to send a whole JS object, rather only the ID or similar.

**Comment by ryanwtyler** [comment]
I really need to be able to map my values to keys.  Like any html select elements

**Comment by ryanwtyler** [comment]
The display value should be the filtered item.  If you can't submit both value and key then the key should be sent with form submission.

**Comment by jounik** [comment]
Right.

As a workaround, you can use objects in the `items` array, and provide a `toString()` method for each object which returns the display value.

For example:

```javascript
var list = [
  { key: 1, value: "First option" },
  { key: 2, value: "Second option" }
];

list.map(function(item) {
  item.toString = function() {
    return this.value;
  };
});

document.querySelector("vaadin-combo-box").items = list;
```

Here’s a Codepen for the same thing: http://codepen.io/jouni/pen/jWeqzM

**Comment by jounik** [comment]
Please do a quick round of DX tests (1–2 people outside the team) to validate that whatever we implement is understood by someone else and is usable.

**Comment by albertolobrano** [comment]
I can test if you like. I am interested in this feature and for the time being i am using  your work around

**Comment by jounik** [comment]
Thanks Alberto, much appreciated!

Currently it seems like this is a bit of a rabbit hole down to item template support, custom filtering, and all things around how the value of the combo box is handled.

Initially we were looking to adding a quick support that would avoid the need for the workaround, and release that in 1.0, but right now I feel like we might want to postpone this a little further so that we can release a stable 1.0 as soon as possible. This object type support would then be added in 1.1, hopefully very soon after 1.0.

**Comment by platosha** [comment]
## Research Report

During this research, the evolved from idea to draft implementation. The feature design questions were thought out and now have answers prototyped. The resulting prototype implementation was evaluated in discussion with @Saulis.

Prototype branch: https://github.com/vaadin/vaadin-combo-box/tree/proto/objects-for-items

Introducing a proper support for using objects as items involves a complex change, even without the item template support. There’re  a lot of places in the current master version that we need to change for that.

### Conclusions

With this feature implemented, the combo-box should have:

- Separate properties for `selectedItem` object and `value` string, to support both using inside a dynamic view and a form.
  - We might want to make both of these properties writeable, changing the selected item on write, there’re use cases for that as well.
    - Writeable `selectedItem` would allow use with bindings and would be consistent with `iron-list`.
    - Having a writeable `value`, at the same time, allows to use a string value to set the selected item, e. g., when initializing the combo-box with attributes.
- `itemLabelPath` and `itemValuePath` properties for pointing to the item keys containing the displayed label and the form value, respectively.
  - Using paths to support nested keys (like: `"user.account.id"`) is easy and makes the implementation more flexible and universal, we should do it.
  - These properties should have default values of `label` and `value`, respectively, to be consistent with `paper-dropdown-menu` and conform  HTML `<option/>` tag attributes.
- `toString()` fallback. If the item object doesn’t contain the `itemLabelPath` or `itemValuePath` key, the `item.toString()` value should be used.

### Custom input values with object items

Supporting custom input values when using object items is another complex question that probably needs to be scoped-out.

It requires to make a callback for providing the combo-box with the item instance constructed after the custom input value. The callback should be invoked either in `_filterChanged` or in `_onClosed` method.

In my prototype, I've just disabled custom input values when using objects as items.

### Custom item template support

Exposing items template is one step beyond the scope of this task. Despite that, it was also prototyped to get the idea of its complexity level.

The prototype branch contains a draft but working implementation: https://github.com/vaadin/vaadin-combo-box/tree/proto/custom-item-template

Exposing an item template is a complex task as well (roughly at the sample level, I’d say). While possible per se, still it requires to introduce a separate `vaadin-combo-box-item` element and expose the `iron-list` in the light DOM of the overlay. Deep explanation of the problems below.

- To achieve that, the first idea was to set the entire `iron-list` item template with the template provided in the combo-box light DOM.
    - Turned out to be impractical: we still need to have an `.item` node defined inside the overlay, assigning model and state attributes (like `id` with index, `selected`, `focused`, ARIA `role` and `aria-selected`) and basic styles to it.
- The second idea was: instead of replacing the entire item template, try to compose the built-in item template with the one from the outside, inserting it manually inside the `.item` wrapper. 
  - Hard to implement. The puzzle is: when registering the overlay element, Polymer parses the contents of `<template>` nodes as well, extracting bindings into the annotations. After this parsing at the register time, it’s impossible to modify the contents of the item template.
    - Tried to disable parsing bindings on the overlay register by adding `<template preserve-content>` (undocumented Polymer feature) attribute to the `iron-list` item template. This makes it possible to modify the item template, but at the same time effectively prevents the `.item` bindings to the overlay methods from being parsed and working. Hence the `focused` indication and ARIA attributes don’t work.
- The working idea: instead of manually modifying the item template, change the `div.item` into a separate `vaadin-combo-box-item` element, that would support custom template in the `template` property. This is currently implemented in the prototype branch.
  - For the item element, we need to `templatize()` it with either a given template or a default item template, stamp a template instance and append it to the light DOM. The changes to the bound `item` model should be updated on the stamped template instance manually.
  - `iron-list` exposes the item nodes in the light DOM. But still the overlay also needs to expose the `iron-list` in the light DOM, in order to be targetable with global CSS.

**Comment by jounik** [comment]
Split the "allow-custom-value" property to a separate PR (#154) which should be merged for 1.0. Custom input values should disabled by default.

**Comment by jounik** [comment]
The layout changes related to custom input values are handled by https://github.com/vaadin/components-team-tasks/issues/82

**Comment by Saulis** [comment]
@albertolobrano if you're interested, you could try out the dev branch we're using and tell us your thoughts about the API. All feedback would be appreciated.

You can checkout the `feature/support-for-object` branch or install using bower:

```bower install vaadin-combo-box#feature/support-for-object```

**Comment by albertolobrano** [comment]
very good. I will give it a try tomorrow and let you know.

**Comment by albertolobrano** [comment]
I tested the component and it is working very well.  I will keep using this in my project until you merge it into master.  exactely what i needed.

**Comment by Saulis** [comment]
@albertolobrano glad you liked it!

Is there anything you would like to change? Anything at all?

**Comment by albertolobrano** [comment]
I am installing this component everywhere so that I can test it better.  Once I will start to use it in real life scenario I will let you know. But is working just fine.
